onnxruntime/core/codegen/target/generic/op_ir_creator/tensor/cast.cc

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/codegen/target/generic/op_ir_creator/all_ops.h"
+
+#include "core/codegen/mti/tensor/cast_ops.h"
+#include "core/codegen/target/ort_tvm_utils.h"
+#include "core/framework/op_kernel_info.h"
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+// Evaluate of Cast OpIRCreator
+Status GENERIC_OP_IR_CREATOR_CLASS(Cast)::Evaluate(
+    const tvm::Array<tvm::Tensor>& inputs,
+    const Node& node,
+    CodeGenContext&,
+    tvm::Array<tvm::Tensor>& outputs) {
+  ProtoHelperNodeContext ctx(node);
+  OpNodeProtoHelper<ProtoHelperNodeContext> attrs(&ctx);
+
+  int64_t to;
+  ORT_RETURN_IF_ERROR(attrs.GetAttr<int64_t>("to", &to));
+  auto to_type_proto = gsl::narrow_cast<ONNX_NAMESPACE::TensorProto_DataType>(to);
+
+  tvm::Tensor X = inputs[0];
+  tvm::Tensor Y;
+  if (to_type_proto == ONNX_NAMESPACE::TensorProto_DataType_BOOL) {
+    // special case for bool as ONNX bool is uint8, while in tvm it's uint1
+    Y = CastToUInt8Bool(X, node.Name() + "_Cast");
+  } else {
+    Y = Cast(X, ToTvmType(to_type_proto), node.Name() + "_Cast");
+  }
+
+  outputs.push_back(Y);
+  return Status::OK();
```

Comment:
so those are the impl to iterate onnx graph and convert to tvm ir, right? do we plan let this one to be the standard interface for onnx graph --> tvm? or it is just nuphar's current implementation? if this is only for nuphar, it is better move to nuphar folder.

Reply:
Didn't get the question. Is this related to cast? or int8 to bool thing?

Reply:
tvm bool and ONNX bool are different types (1-bit vs. 8-bit), so this is general for ONNX to tvm

---
In reply to: [294448930](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294448930) [](ancestors = 294448930)

Reply:
Hmm... it looks codeflow has bug, it put my question to the wrong places..... :(
I just want to have a standard way to iterate the onnx graph with a visitor pattern. anyway, we could discuss it later.

---
In reply to: [296499535](https://github.com/microsoft/onnxruntime/pull/888#discussion_r296499535) [](ancestors = 296499535,294448930)

Reply:
forget about it, codeflow bug, i guess

---
In reply to: [294552124](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294552124) [](ancestors = 294552124)

Reply:
We has a way to iterate onnx graph, but we just don't know whether it is representative enough to be a "standard". That is why we move it to nuphar. If it turns out to be common enough, we can move it back.